src/tilemap_layer.cpp

```diff
@@ -141,35 +142,66 @@ TilemapLayer::TilemapLayer(int ilayer) :
 	substitutions(ilayer >= 1
 			? Main_Data::game_data.map_info.upper_tiles
 			: Main_Data::game_data.map_info.lower_tiles),
-	layer(ilayer) {
-
-	memset(autotiles_ab, 0, sizeof(autotiles_ab));
-	memset(autotiles_d, 0, sizeof(autotiles_d));
-
-	// SubLayer for the tiles with Wall or Above passability
-	// Its z-value should be between the z of the events in the upper layer and the hero
-	sublayers.push_back(std::make_shared<TilemapSubLayer>(this, Priority_TilesetAbove + layer));
+	layer(ilayer),
 	// SubLayer for the tiles without Wall or Above passability
 	// Its z-value should be under z of the events in the lower layer
-	sublayers.push_back(std::make_shared<TilemapSubLayer>(this, Priority_TilesetBelow + layer));
+	lower_layer(this, Priority_TilesetBelow + layer),
+	// SubLayer for the tiles with Wall or Above passability
+	// Its z-value should be between the z of the events in the upper layer and the hero
+	upper_layer(this, Priority_TilesetAbove + layer)
+{
 }
 
-void TilemapLayer::DrawTile(Bitmap& dst, Bitmap& screen, int x, int y, int row, int col, bool allow_fast_blit) {
-	Bitmap::TileOpacity op = screen.GetTileOpacity(row, col);
+EP_ALWAYS_INLINE
+void TilemapLayer::DrawTile(Bitmap& dst, Bitmap& tileset, Bitmap& tone_tileset, int x, int y, int row, int col, uint32_t tone_hash, bool allow_fast_blit) {
+	auto op = tileset.GetTileOpacity(col, row);
+	if (op != ImageOpacity::Transparent) {
+		DrawTileImpl(dst, tileset, tone_tileset, x, y, row, col, tone_hash, op, allow_fast_blit);
```

Comment:
I don't understand why there is now a DrawTileImpl function call. Besides inverting the if why isn't the cost just put in here?

Reply:
The 2 level functions here is to inline the transparency check. For the common case (especially on upper level) where we have transparent tiles avoid paying for a function call just to do the check and return.

This also makes the check duplicated in `Draw()` for every tile type instead of just existing in one place in `DrawTile()`. Now that there are separate branches, that allows the branch predictor to predict separate against each tile type, instead of having one prediction for the entire tile set. Since each class of tiles has a different probability of being transparent, this is a net win.

Both of my vtune profile examples showed a gain of 1-2% less time spent drawing tiles with this change,

Reply:
okay, can you add a short comment there that this is force-inline + call is intentional as it is?

To prevent others from killing this optimisation later by accident..

Reply:
Sounds good, added